sensors/baro/ms5611.c

```diff
@@ -0,0 +1,280 @@
+
+/* Read ADC conversion result command */
+#define CMD_READ_ADC 0x00
+
+/* Read PROM command and PROM address specifiers */
+#define CMD_READ_PROM 0xa0
+#define PROM_SIZE     8
+
+/* MS5611 DELAYS */
+
+/* OSR dependent minimum wait (microseconds) for conversion. Datasheet values ceil-ed to full milliseconds */
+#define OSR_256_SLEEP  1000
+#define OSR_512_SLEEP  2000
+#define OSR_1024_SLEEP 3000
+#define OSR_2048_SLEEP 5000
+#define OSR_4096_SLEEP 10000
+
+/* Device reset duration */
+#define RESET_SLEEP 3000 /* Datasheet says 2800. It seems some data may be lost with such sleep */
+
+
+typedef struct {
+	spimsg_ctx_t spiCtx;
+	oid_t spiSS;
+	sensor_event_t evtBaro;
+	uint16_t prom[PROM_SIZE]; /* Prom memory: [ reserved, C1, C2, C3, C4, C5, C6, CRC ] */
+	char stack[512] __attribute__((aligned(8)));
+} ms5611_ctx_t;
+
+
+/* Based on sample code in AN520 application note for MS56xx/MS57xx/MS58xx pressure sensor by Measurement Specialties */
+uint8_t ms5611_crc4(const uint16_t *prom)
+{
+	uint16_t cnt, bit, word;
+	uint32_t rem = 0; /* crc reminder */
+
+	/* Operation is performed on bytes */
+	for (cnt = 0; cnt < (PROM_SIZE * 2); cnt++) {
+		/* The CRC code is calculated and written in factory with the LSB byte in the prom (prom[7]) set to 0x00 */
+		word = (cnt == (PROM_SIZE * 2 - 1)) ? 0 : prom[cnt >> 1];
+
+		/* High/low byte selection */
+		if (cnt & 1 == 1) {
+			word <<= 8;
+		}
+		rem ^= word >> 8;
+
+		for (bit = 8; bit > 0; bit--) {
+			rem = ((rem & 0x8000) > 0) ? (rem << 1) ^ 0x3000 : (rem << 1);
+		}
+	}
+
+	rem = (0x000f & (rem >> 12)); /* final 4-bit reminder is CRC code */
+
+	return (rem ^ 0x00);
+}
+
+
+/* Sends conversion request `convCmd`, waits `usDelat` for conversion and reads ADC to `res` */
+static int ms5611_measure(ms5611_ctx_t *ctx, uint8_t convCmd, time_t usDelay, uint32_t *res)
+{
+	static const uint8_t readCmd = CMD_READ_ADC;
+	uint8_t val[3];
+
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &convCmd, sizeof(convCmd), NULL, 0, 0) < 0) {
+		fprintf(stderr, "ms5611: failed conv. request\n");
+		return -1;
+	}
+
+	usleep(usDelay);
+
+	if (sensorsspi_xfer(&ctx->spiCtx, &ctx->spiSS, &readCmd, sizeof(readCmd), val, sizeof(val), sizeof(readCmd)) < 0) {
+		fprintf(stderr, "ms5611: adc read failed\n");
+		return -1;
+	}
+
+	/* If ADC is read too early then it returns 0 */
+	if (val[0] == 0 && val[1] == 0 && val[2] == 0) {
+		return -1;
+	}
+
+	*res = ((uint32_t)val[0] << 16) | ((uint32_t)val[1] << 8) | ((uint32_t)val[2]);
+
+	return 0;
+}
+
+
+static void ms5611_publishthr(void *data)
+{
+	sensor_info_t *info = (sensor_info_t *)data;
+	ms5611_ctx_t *ctx = info->ctx;
+	int32_t temp, press;
+	time_t now;
+
+	/* breaking naming convention on purpose to match datasheet notation 1:1 */
+	uint32_t D1 = 0, D2 = 0;
+	int32_t dT;
+	int64_t off, sens;
+
+	/* Initialize temperature with too big delay for certain response */
+	while (ms5611_measure(ctx, (CMD_CVRT_TEMP | CVRT_OSR_4096), OSR_4096_SLEEP * 2, &D2) < 0)
```

Comment:
These two usages of ms5611_measure (here and next usage) could result in infinite loop on HW error, not a merge blocker imho, but would be nice to implement timeout